yellow-paper/docs/calls/public_private_messaging.md

```diff
@@ -0,0 +1,52 @@
+
+The figure below shows the order of function calls on the left-hand side, while the right-hand side shows how the functions will be executed. Notably, the second private function call is independent of the output of the public function and merely occurs after its execution.
+
+![Public - Private Ordering](./images/calls/pvt_pub_ordering.png)
+
+## Private -> Public Messaging
+If a private function in an Aztec smart contract wants to call a public function, it gets pushed into a separate public call stack that is enqueued. The private kernel circuit which must prove the execution of the private function(s), then hashes each of the item in the call stack and returns that. The private kernel proof, the public inputs of the private kernel (which contain the hash of the each of the public call stack item) and other transaction data (like enqueued public function calls, new commitments, nullifiers etc) get passed along to the sequencer. Sequencer then picks up the public call stack item and executes each of the functions. The Public VM which executes the methods then verifies that the hash provided by the private kernel matches the current call stack item.
+
+This way, you can destroy your private state and create them in public within the same transaction or indirectly assert constraints on the execution of the private functions with latest data.
+
+### Handling Privacy Leakage and `msg.sender`
+In the above design, the sequencer only sees the public part of the call stack along with any new commitments, nullifiers etc that were created in the private transaction i.e. should learns nothing more of the private transaction (such as its origin, execution logic etc).
+
+But what if the enqueued public function makes use of `msg_sender` which is meant to use 
+
+Specifically, when the call stack is passed to the kernel circuit, the kernel should assert the `msg_sender` is 0 and hash appropriately. `msg_sender` could be the contract address too instead of `0`, but it leaks which contract is calling the public method and therefore leaks which contract the user was interacting with in private land.
+
+### Reverts
+
+If the private part of the transaction reverts, then public calls are never enqueued. But if the public part of the transaction reverts, it should still revert the entire transaction i.e. the sequencer should drop the execution results of the private part of the transaction and not include those in the state transitioner smart contract. However, since the sequencer had to execute your transaction, appropriate fee will be charged. Reverting in public causing the whole transaction to be dropped enables existing paradigms of ethereum where your valid transaction can revert because of altered state e.g., trade incurring too much slippage.
```

Comment:
It should drop the state changes (effects of the tx), guess that is the same as the execution result, but for result I'm often thinking the return values etc.

Also, we are not including it in the contract, but in the block input to the contract. 

A note here on reverts, I still think the first nullifier is desired to ensure that the tx cannot just be to drain fees. So there is still some effect of it, kinda like the nonce increasing 🤷.